src/crypto/chachapoly_aead.cpp

```diff
@@ -0,0 +1,90 @@
+#include <crypto/chachapoly_aead.h>
+
+#define __STDC_WANT_LIB_EXT1__ 1
+#include <crypto/poly1305.h>
+#include <string.h>
+
+#include <compat/endian.h>
+#include <support/cleanse.h>
+
+#ifndef HAVE_TIMINGSAFE_BCMP
+
+int timingsafe_bcmp(const unsigned char *b1, const unsigned char *b2, size_t n) {
+  const unsigned char *p1 = b1, *p2 = b2;
+  int ret = 0;
+
+  for (; n > 0; n--)
+    ret |= *p1++ ^ *p2++;
+  return (ret != 0);
+}
+
+#endif /* TIMINGSAFE_BCMP */
+
+int chacha20poly1305_init(struct chachapolyaead_ctx *ctx, const uint8_t *key,
+                          int keylen) {
+  if (keylen != (32 + 32)) /* 2 x 256 bit keys */
+    return -1;
+  chacha_keysetup(&ctx->main_ctx, key, 256);
+  chacha_keysetup(&ctx->header_ctx, key + 32, 256);
+  return 0;
+}
+
+int chacha20poly1305_crypt(struct chachapolyaead_ctx *ctx, uint32_t seqnr,
+                           uint8_t *dest, const uint8_t *src, uint32_t len,
+                           uint32_t aadlen, int is_encrypt) {
+  const uint8_t one[8] = {1, 0, 0, 0, 0, 0, 0, 0}; /* NB little-endian */
+  uint8_t expected_tag[POLY1305_TAGLEN], poly_key[POLY1305_KEYLEN];
+  int r = -1;
+
+  uint64_t seqnr64 = seqnr;
+  uint64_t chacha_iv = htole64(seqnr64);
+  memset(poly_key, 0, sizeof(poly_key));
+  chacha_ivsetup(&ctx->main_ctx, (uint8_t *)&chacha_iv, NULL);
```

Comment:
Use `nullptr` :-)

Reply:
I'd like to keep this as C code for now, the .cpp filename extension is for bypassing the CXXFLAGS/CFLAGS issue.